Apply network resource Network

zaibon opened the pull request:
fixes #61

delandtj committed:
testsetup fixes

zaibon committed:
Logic to apply network resource

fixes #61

zaibon committed:
fix bridge package tests

delandtj commented on modules/network/network_test.go:
```diff
@@ -46,13 +57,94 @@ func TestCreateBridge(t *testing.T) {
+							Peer:     net.ParseIP("2001::1"),
+							PeerPort: 51820,
+							Key:      "4w4woC+AuDUAaRipT49M8SmTkzERps3xA5i0BW4hPiw=",
+							// LinkLocal: net.
```
fe80::cc01/64 to add for testing ?

delandtj commented on modules/network/networker.go:
```diff
@@ -1,27 +1,200 @@
+		err         error
+	)
+
+	key, err = wireguard.LoadKey(storagePath)
```
what if the error is a storage error ?

zaibon replied:
Well in any case, If I cannot read the key from the disk, would it be because it has not been generated before or storage error, I need to create a new one.
But this means I need a way to update the resource unit when things like that happens, cause all the other node will have to upgrade their wg config with the new public key.

delandtj commented on modules/network/networker.go:
```diff
@@ -1,27 +1,200 @@
+		peers[i] = wireguard.Peer{
+			PublicKey:  conn.Key,
+			Endpoint:   endpoint,
+			AllowedIPs: []string{"0.0.0.0/0"},
```
AllowedIPs has to contain: the fe80::/ of the WG, the Prefix of the Network Resource on the bridge, an IPv4 LL addr and an IPv4 subnet on the bridge. No peer can heve the same IP (so certainly not 0/0)

zaibon replied:
But don't you need this for the wg interface to be able to forward all traffic in/out ? I though that without this the interface can actually only reach other peer. But the exit not cannot send public traffic to it

delandtj replied:
Hmm.. 
I'll setup testnet.sh to reflect that properly

delandtj commented on modules/network/wireguard/wireguard.go:
```diff
@@ -141,3 +163,28 @@ func newPeer(pubkey, endpoint string, allowedIPs []string) (wgtypes.PeerConfig,
 	return peer, nil
 }
+
+func GenerateKey(dir string) (wgtypes.Key, error) {
```
hmmm... you have multiple error types emanating from that function, how to handle them ?

zaibon replied:
Well it up to the caller to decide what to do when something fails.

zaibon committed:
Logic to apply network resource

delandtj committed:
Routing/firewalling specs

zaibon committed:
address lint warnings

delandtj committed:
Progress

delandtj committed:
forreal now

delandtj committed:
setupwgtest.sh and User Network Object

  - added specifications for reachable peers in network.go
  - added a container behind NRs to verify correct routing/
  forwarding.

delandtj committed:
Syntax errors

zaibon committed:
Update network.go

zaibon committed:
add network/ip package

ip package contain logic to generate deteminictic names based on an IP

zaibon committed:
remove unused code

zaibon committed:
fix some tests and add doc strings

zaibon committed:
add DeleteNetResource to networker interface

zaibon committed:
fix nibble tests

zaibon committed:
refactor veth address configuration
